keyberon firmware

fabi1cazenave opened the pull request:
Taken from the [nibble-rp2040-rs](https://github.com/DrewTChrist/nibble-rp2040-rs), which is based on an Adafruit rb2040, which should be close enough to the ProMicro. (Worst case, we could propose an Adafruit controller for the first keebs.)

Not working yet: Nibble’s code looks outdated, the rp2040-hal in particular has come a long way, and it doesn’t build as is — the USB abstraction doesn’t match keyberon’s expectation.

fabi1cazenave committed:
(wip) keyberon firmware

fabi1cazenave committed:
credentials

Nuclear-Squid committed:
Updated `.gitignore` file.

Nuclear-Squid commented:
I didn’t check your code yet but just so you know, the `target` folder is the output of the Rust compiler and should be git-ignored. I’m sorry, I’m just not gonna read through a PR with >1k files changed :sweat_smile: 
<img width="196" height="79" alt="Screenshot_2025-07-29_20-02-38" src="https://github.com/user-attachments/assets/690ed1b8-141d-4cfe-94db-0d8a43ea80fd" />
I updated the `.gitignore` file and removed the `target` folder.

fabi1cazenave commented:
yes, I’ve made a mistake with my last commit.

I’m not ***THAT*** dumb.

Nuclear-Squid committed:
Fixed compilation errors related to usb.

fabi1cazenave committed:
credentials

Nuclear-Squid commented:
Hey I’m really sorry for insulting you like that. You said earlier in the day that it was a your first serious Rust project and asked a couple questions about the file hierarchy (what the `rust.raw` file was, why `Cargo.lock` was so heavy…), so I *genuinely* thought you might not know.

My comment was very badly worded (especially the "I’m not gonna read a PR with 1k files changed"), and came off as an asshole when really all I wanted to say was "Hey, in case you didn’t know, `.gitignore` should be setup that way, I fixed it real quick".

It’s fine to make mistakes, I would never intentionally be mean to you for a mistake I can fix in 5 minutes. I should have checked the wording more carefully (or even just not say anything at all honestly), and I’m terribly sorry for this mess that could have been avoided if I just was more careful of what I said.

PS : I fixed the compilation errors related to the USB bus. It still does not compile, since the HAL needs to be updated, but I’ll see what I can do.

fabi1cazenave commented:
No worries.

fabi1cazenave commented:
In the Keymini implementation (stm32), and as far as I understand, the USB magic comes from the STM32 hal:

```rust
use hal::usb;
// ...;
use stm32f0xx_hal as hal;
use usb_device::bus::UsbBusAllocator;
use usb_device::class::UsbClass as _;
use usb_device::device::UsbDeviceState;

type UsbClass = keyberon::Class<'static, usb::UsbBusType, ()>;
type UsbDevice = usb_device::device::UsbDevice<'static, usb::UsbBusType>;
```

which is used as is in the app:

```rust
    #[shared]
    struct Shared {
        usb_dev: UsbDevice,
        usb_class: UsbClass,
    }
```

A possible rp2040 equivalent for `UsbClass` *could* be the following:

```rust
use rp2040_hal::usb as usb;

type UsbClass = keyberon::Class<'static, usb::UsbBus, ()>;
```

But I’m still confused by `UsbDevice`.

fabi1cazenave commented:
Oh. Nevermind, that was in your commit. So it’s been a beautiful waste of time for both of us.

Nuclear-Squid commented:
It’s not a waste of time, while the final changes weren’t that big, it was surprisingly tricky to get it to work. We’re now one step closer to have a firmware for the keyboard, which is very important.

fabi1cazenave commented:
Sorry for the time loss. Next time I’ll make sure my PR is in an acceptable state before submitting it for comments.

fabi1cazenave commented:
TIL: all chip-specific crates have their own API (`stm32_hal`, `rp2040_hal`…), but that’s not an issue as long as they implement the `embedded_hal` traits that Keyberon relies on. Neat.

It took me a while, but I think that’s how the current `rp2040_hal` API could be used to declare a key matrix:

```rust
    // rp2040 implementations of the embedded_hal::digital::{InputPin,OutputPin} traits
    type InputPin  = gpio::Pin<gpio::DynPinId, gpio::FunctionSioInput,  gpio::PullUp>;
    type OutputPin = gpio::Pin<gpio::DynPinId, gpio::FunctionSioOutput, gpio::PullUp>;

    #[shared]
    struct Shared {
        usb_dev: UsbDevice,
        usb_class: UsbClass,
        timer: Timer,
        alarm: Alarm0,
        #[lock_free]
        matrix: Matrix<InputPin, OutputPin, 12, 4>,
        layout: Layout<12, 4, 1, kb_layout::CustomActions>,
        #[lock_free]
        debouncer: Debouncer<[[bool; 16]; 5]>,
        #[lock_free]
        watchdog: Watchdog,
    }
```

I’ll send a PR a bit later. Still working on it.

fabi1cazenave committed:
wip: switch to the latest rp2040_hal API

fabi1cazenave commented:
This commit *looks* like the `InputPin` and `OutputPin` types are defined properly, but there’s a catch I don’t get when declaring the Keyberon Matrix elements:

```rust
        let matrix = Matrix::new(
            [
                pins.gpio3.into_pull_up_input(),
                pins.gpio4.into_pull_up_input(),
```

The `gpio3` line passes, but the `gpio4` line gets me this error:

```
error[E0308]: mismatched types
   --> src/main.rs:152:17
    |
152 |                 pins.gpio4.into_pull_up_input(),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Pin<Gpio3, FunctionSio<SioInput>, PullUp>`, found `Pin<Gpio4, FunctionSio<SioInput>, PullUp>`
    |
    = note: expected struct `rp2040_hal::gpio::Pin<Gpio3, _, _>`
               found struct `rp2040_hal::gpio::Pin<Gpio4, _, _>`
```

Any idea?

fabi1cazenave committed:
12*4 => 8*6 grid

fabi1cazenave committed:
low-hanging fruits (warnings)

fabi1cazenave committed:
proper type casting for gpio pins

fabi1cazenave committed:
proper build target

fabi1cazenave committed:
basic keyboard layout

Nuclear-Squid committed:
Fixed linking error: necessary features on crate `rp2040-hal`.

Nuclear-Squid committed:
Fixed firmware upload script.

Nuclear-Squid committed:
RP2040 chip is now recognised as a USB peripheral.

Nuclear-Squid committed:
Cleanup

Nuclear-Squid committed:
Usb stuck in `Suspend` mode, pls help.

Nuclear-Squid committed:
Firmware very close to Keymini’s but no key presses ar ever detected.

fabi1cazenave committed:
comment nitpicks

Nuclear-Squid committed:
Wip (matrix).

fabi1cazenave committed:
transpose the Keyberon matrix

fabi1cazenave committed:
add a QuackenZero-specific matrix

fabi1cazenave committed:
fixes: proper pin refs + 1 microsecond delay

fabi1cazenave commented:
This works nicely. Now I have to include a good default layout, and this should be ready to ship. ^^

fabi1cazenave committed:
proper configuration

fabi1cazenave committed:
drop unused rp2040-hal features

fabi1cazenave commented:
@Nuclear-Squid this should be ready for a review if you have time. :-)

fabi1cazenave committed:
nitpick

Nuclear-Squid committed:
Swapped `cortex-m-rtic` for `rtic`, due to sub-dependencies issues.

`cortex-m-rtic` depends on `heapless-0.7.7`, which in turn depends
on `atomic-polyfill-1.0.0`, a deprecated crate that just **doesn’t**
compile on my machine, and shouldn’t even be used in the first place.

Swappinging it out for `rtic` solves this issue without affecting
how the keyboard’s work at run-time.

fabi1cazenave committed:
oops, forgot the README

fabi1cazenave committed:
oops

Nuclear-Squid commented:
Looks good to me !
I have some *very minor nitpicks*, but it’s just syntax and not even really worth discussing here, so I’m down to merge this.

I just pushed a commit to swap out some dependencies, since it just would not compile on my machine (a deprecated crate caused issues). I added context in the commit body. Make sure it still compiles on your end, just in case. If it does, I’ll merge it

fabi1cazenave commented:
I saw that, good catch. We might want to document it somehow.

fabi1cazenave committed:
a bit of doc

fabi1cazenave committed:
Switching to 2024 (stable).

The recent switch to rtic[thumbv6] broke the build on my box.
Upgrading to edition2024 and the latest stable Cargo seems to fix it.
